File: src/main/java/com/github/wgzhao/dbquery/controller/admin/QueryConfigController.java
package com.github.wgzhao.dbquery.controller.admin;

import com.github.wgzhao.dbquery.dto.CommResponse;
import com.github.wgzhao.dbquery.dto.DbSourceDto;
import com.github.wgzhao.dbquery.dto.QueryResult;
import com.github.wgzhao.dbquery.entities.DataSources;
import com.github.wgzhao.dbquery.entities.QueryConfig;
import com.github.wgzhao.dbquery.entities.QueryParam;
import com.github.wgzhao.dbquery.repo.DataSourceRepo;
import com.github.wgzhao.dbquery.repo.QueryConfigRepo;
import com.github.wgzhao.dbquery.repo.QueryParamRepo;
import com.github.wgzhao.dbquery.service.ConnectionDB;
import com.github.wgzhao.dbquery.util.CacheUtil;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.sql.SQLException;
import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * QueryConfigController
 * manipulate for {@link QueryConfig} table
 */
@RestController
@RequestMapping("${app.api.manage-prefix}/queryConfig")
@Slf4j
public class QueryConfigController {

    @Autowired
    private QueryConfigRepo queryConfigRepo;

    @Autowired
    private DataSourceRepo dataSourceRepo;

    @Autowired
    private QueryParamRepo queryParamsRepo;

    @Resource
    private CacheUtil cacheUtil;

    @Autowired
    private ConnectionDB connectionDB;

    /**
     * list all query config
     * @return list of {@link QueryConfig}
     */
    @GetMapping
    public List<QueryConfig> list() {
        return queryConfigRepo.findAll();
    }

    @GetMapping("/{id}")
    public QueryConfig get(@PathVariable("id") String id) {
        return queryConfigRepo.findById(id).orElse(null);
    }

    @DeleteMapping("/{id}")
    public void delete(@PathVariable("id") String id) {
        queryConfigRepo.deleteById(id);
    }

    @GetMapping("/datasources")
    public Collection<DbSourceDto> listDataSources() {
        return dataSourceRepo.findNoAndName();
    }

    @PostMapping
    public QueryConfig save(@RequestBody  QueryConfig queryConfig) {
        return queryConfigRepo.save(queryConfig);
    }

    @GetMapping("/params/{selectId}")
    public List<QueryParam> listParams(@PathVariable("selectId") String selectId) {
        return queryParamsRepo.findBySelectId(selectId);
    }

    @PutMapping("/params")
    public int saveParams(@RequestBody  List<QueryParam> params) {
        queryParamsRepo.saveAll(params);
        return params.size();
    }

    @DeleteMapping("/cache/{selectId}")
    public int deleteCache(@PathVariable("selectId") String selectId) {
        return cacheUtil.deleteKeys(selectId);
    }

    @PostMapping("/testQuery")
    public QueryResult testQuery(@RequestBody Map<String, String> payload) {
        String sourceId = payload.get("sourceId");
        String querySql = payload.get("querySql");
        DataSources dataSource = dataSourceRepo.findById(sourceId).orElse(null);
        QueryResult queryResult = new QueryResult();
        if (dataSource == null) {
            queryResult.setStatus(400);
            queryResult.setMessage("数据源不存在");
            return queryResult;
        }
        try {
            List<Map<String, Object>> rsList = connectionDB.executeSQL(querySql, dataSource);
            queryResult.setStatus(200);
            queryResult.setTotal(rsList.size());
            queryResult.setData(Map.of("result", rsList));
            return queryResult;
        } catch (SQLException | ClassNotFoundException e) {
            log.error("test query failed: {}", e.getMessage());
            queryResult.setStatus(400);
            queryResult.setMessage(e.getMessage());
            return  queryResult;
        }
    }
}

File: src/main/kotlin/com/github/wgzhao/dbquery/controller/admin/QueryConfigController.kt
package com.github.wgzhao.dbquery.controller.admin

import com.github.wgzhao.dbquery.dto.DbSourceDto
import com.github.wgzhao.dbquery.dto.QueryResult
import com.github.wgzhao.dbquery.entities.QueryConfig
import com.github.wgzhao.dbquery.service.ConnectionDB
import com.github.wgzhao.dbquery.service.DataSourceService
import com.github.wgzhao.dbquery.service.QueryConfigService
import com.github.wgzhao.dbquery.util.CacheUtil
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.web.bind.annotation.*
import java.sql.SQLException

/**
 * QueryConfigController
 * manipulate for [QueryConfig] table
 */
@RestController
@RequestMapping("\${app.api.manage-prefix}/query-configs")
class QueryConfigController(
    val queryConfigService: QueryConfigService,
    val dataSourceService: DataSourceService,
    val cacheUtil: CacheUtil,
    val connectionDB: ConnectionDB
) {
    private val logger = KotlinLogging.logger {}

    /**
     * list all query config
     * @return list of [QueryConfig]
     */
    @GetMapping
    fun list(): MutableList<QueryConfig?> {
        return queryConfigService.getAll()
    }

    @GetMapping("/{id}")
    fun get(@PathVariable("id") id: String): QueryConfig? {
        return queryConfigService.findById(id).orElse(null)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable("id") id: String) {
        queryConfigService.deleteById(id)
    }

    @GetMapping("/data-sources")
    fun listDataSources(): Collection<DbSourceDto> {
        return dataSourceService.findNoAndName()
    }

    @PostMapping
    fun save(@RequestBody queryConfig: QueryConfig): QueryConfig {
        return queryConfigService.save(queryConfig)
    }

    @DeleteMapping("/cache/{selectId}")
    fun deleteCache(@PathVariable("selectId") selectId: String?): Int {
        return cacheUtil.deleteKeys(selectId)
    }

    @PostMapping("/testQuery")
    fun testQuery(@RequestBody payload: MutableMap<String?, String>): QueryResult {
        val sourceId: String = payload["sourceId"]!!
        val querySql = payload["querySql"]
        val dataSource =
            dataSourceService.findById(sourceId).orElse(null) ?: return QueryResult.error(400, "数据源不存在")

        try {
            val rsList = connectionDB.executeSQL(querySql, dataSource)
            // normalize to List<Map<String, Any>>
            val normalized: List<Map<String, Any>> = rsList.mapNotNull { row ->
                row?.entries?.mapNotNull { (k, v) -> k?.let { it to (v as Any) } }?.toMap()
            }
            return QueryResult.success(normalized)
        } catch (e: SQLException) {
            logger.error { "${"test query failed: {}"} ${e.message}" }
            return QueryResult.error(400, e.message)
        } catch (e: ClassNotFoundException) {
            logger.error { "${"test query failed: {}"} ${e.message}" }
            return QueryResult.error(400, e.message)
        }
    }
}
